source/endpoints/ruleset.c

```diff
@@ -0,0 +1,878 @@
+/**
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+#include <aws/common/array_list.h>
+#include <aws/common/hash_table.h>
+#include <aws/common/json.h>
+#include <aws/common/ref_count.h>
+#include <aws/common/string.h>
+#include <aws/sdkutils/endpoints/private/ruleset_types_impl.h>
+
+#define SKIP_DOC true
+
+/* parameter types */
+static struct aws_byte_cursor s_string_type_cur = AWS_BYTE_CUR_INIT_FROM_STRING_LITERAL("string");
+static struct aws_byte_cursor s_boolean_type_cur = AWS_BYTE_CUR_INIT_FROM_STRING_LITERAL("boolean");
+
+/* rule types */
+static struct aws_byte_cursor s_endpoint_type_cur = AWS_BYTE_CUR_INIT_FROM_STRING_LITERAL("endpoint");
+static struct aws_byte_cursor s_error_type_cur = AWS_BYTE_CUR_INIT_FROM_STRING_LITERAL("error");
+static struct aws_byte_cursor s_tree_type_cur = AWS_BYTE_CUR_INIT_FROM_STRING_LITERAL("tree");
+
+/* TODO: improve error messages. Include json line num? or dump json node? */
+
+/*
+******************************
+* Parameter Getters.
+******************************
+*/
+enum aws_endpoints_parameter_value_type aws_endpoints_parameter_get_value_type(
+    const struct aws_endpoints_parameter *parameter) {
+    AWS_PRECONDITION(parameter);
+    return parameter->type;
+}
+
+const struct aws_string *aws_endpoints_parameter_get_built_in(const struct aws_endpoints_parameter *parameter) {
+    AWS_PRECONDITION(parameter);
+    return parameter->built_in;
+}
+
+int aws_endpoints_parameter_get_default_string(
+    const struct aws_endpoints_parameter *parameter,
+    struct aws_string **out_string) {
+    AWS_PRECONDITION(parameter);
+    AWS_PRECONDITION(out_string);
+
+    if (parameter->type == AWS_ENDPOINTS_PARAMETER_STRING) {
+        *out_string = parameter->default_value.string;
+        return AWS_OP_SUCCESS;
+    };
+
+    return AWS_OP_ERR;
```

Comment:
need to aws_raise_error(SOMETHING)
you can't just `return AWS_OP_ERR` unless some other function has failed and raise an error

likewise for all these getters